numpy/_core/src/multiarray/array_converter.c

```diff
@@ -0,0 +1,563 @@
+                    &item->scalar_input);
+            if (item->array == NULL) {
+                goto fail;
+            }
+        }
+        if (!item->scalar_input) {
+            self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+        }
+        /* At this point, assume cleanup should happen for this item */
+        self->narrs++;
+        Py_INCREF(item->object);
+        item->DType = NPY_DTYPE(PyArray_DESCR(item->array));
+        Py_INCREF(item->DType);
+
+        if (npy_mark_tmp_array_if_pyscalar(
+                item->object, item->array, &item->DType)) {
+            /* just clear the flag again */
+            ((PyArrayObject_fields *)(item->array))->flags &= (
+                    ~NPY_ARRAY_WAS_PYTHON_LITERAL);
+            item->descr = NULL;
+        }
+        else {
+            self->flags &= ~NPY_CH_ALL_PYSCALARS;
+
+            item->descr = PyArray_DESCR(item->array);
+            Py_INCREF(item->descr);
+        }
+    }
+
+    return (PyObject *)self;
+
+  fail:
+    Py_DECREF(self);
+    return NULL;
+}
+
+
+static PyObject *
+array_converter_get_scalar_input(PyArrayArrayConverterObject *self)
+{
+    PyObject *ret = PyTuple_New(self->narrs);
+    if (ret == NULL) {
+        return NULL;
+    }
+
+    creation_item *item = self->items;
+    for (int i = 0; i < self->narrs; i++, item++) {
+        if (item->scalar_input) {
+            Py_INCREF(Py_True);
+            PyTuple_SET_ITEM(ret, i, Py_True);
+        }
+        else {
+            Py_INCREF(Py_False);
+            PyTuple_SET_ITEM(ret, i, Py_False);
+        }
+    }
+    return ret;
+}
+
+
+static int
+find_wrap(PyArrayArrayConverterObject *self)
+{
+    PyObject *wrap = NULL;
+    PyObject *wrap_type = NULL;
+    double priority = NPY_NAN;  /* silence compiler warnings */
+
+    if (self->wrap != NULL) {
+        return 0;  /* nothing to do */
+    }
+
+    for (int i = 0; i < self->narrs; i++) {
+        PyObject *obj = self->items[i].object;
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
+                Py_INCREF(Py_None);
+                Py_XSETREF(wrap, Py_None);
+                priority = 0;
+            }
+        }
+        else if (PyArray_IsAnyScalar(obj)) {
+            if (wrap == NULL || priority < NPY_SCALAR_PRIORITY) {
+                Py_INCREF(Py_None);
+                Py_XSETREF(wrap, Py_None);
+                priority = NPY_SCALAR_PRIORITY;
+            }
+        }
+        else {
+            PyObject *new_wrap = PyArray_LookupSpecial_OnInstance(
+                    obj, npy_ma_str_array_wrap);
+            if (new_wrap == NULL) {
+                if (PyErr_Occurred()) {
+                    goto fail;
+                }
+                continue;
+            }
+            double curr_priority = PyArray_GetPriority(obj, 0);
+            if (wrap == NULL || priority < curr_priority) {
+                Py_XSETREF(wrap, new_wrap);
```

Comment:
You can't get here again after the first time you enter this block. Maybe `break` at the end to enforce that we don't need to do further loop iterations?

Reply:
I think you can get here if the priority of the next input is higher than what was there before.

Reply:
Argh, misread, you're right.